api/envoy/service/ratelimit/v3/rls.proto

```diff
@@ -144,4 +156,16 @@ message RateLimitResponse {
   // - :ref:`envoy.filters.network.ratelimit <config_network_filters_ratelimit_dynamic_metadata>` for network filter.
   // - :ref:`envoy.filters.thrift.rate_limit <config_thrift_filters_rate_limit_dynamic_metadata>` for Thrift filter.
   google.protobuf.Struct dynamic_metadata = 6;
+
+  // Quota granted for the combined descriptor sets. This is a certain number of requests over a
+  // period of time. The client may cache this result and apply the effective RateLimitResponse to
+  // future matching requests with the same descriptor set.
+  //
+  // If quota is available, a RLS request will not be made and the quota will be reduced by 1.
+  //
+  // If there is not sufficient quota, a RLS request will be made. When the result is returned,
```

Comment:
Doesn't this mean that every Nth request will have higher latency due to requiring a round-trip with the RLS server?  It seems like we should try to set things up such that the client already has the next quota before it actually needs it.

One option here would be some sort of pre-fetching mechanism, where the client can fetch a new quota when the old quota reaches some minimum number of remaining requests.

Another option would be a streaming API, where the server could update the client's quota whenever it changes without the client needing to explicitly make a request.  In fact, couldn't the rate limit quota just be another xDS resource type that the client subscribes to?

Reply:
I think the client can optionally make a preemptive RLS request before quota expires. So, this is actually at the discretion of implementation.